runtime/modules/ui.lua

```diff
@@ -0,0 +1,81 @@
+
+-- vim.ui.select({items}, {opts}, {on_choice})                  *vim.ui.select()*
+--     Prompts the user to pick from a list of items, allowing arbitrary
+--     (potentially asynchronous) work until `on_choice`.
+
+--     Example: >lua
+--         vim.ui.select({ 'tabs', 'spaces' }, {
+--             prompt = 'Select tabs or spaces:',
+--             format_item = function(item)
+--                 return "I'd like to choose " .. item
+--             end,
+--         }, function(choice)
+--             if choice == 'spaces' then
+--                 vim.o.expandtab = true
+--             else
+--                 vim.o.expandtab = false
+--             end
+--         end)
+-- <
+--     Parameters: ~
+--       • {items}      (table) Arbitrary items
+--       • {opts}       (table) Additional options
+--                      • prompt (string|nil) Text of the prompt. Defaults to
+--                        `Select one of:`
+--                      • format_item (function item -> text) Function to format
+--                        an individual item from `items`. Defaults to
+--                        `tostring`.
+--                      • kind (string|nil) Arbitrary hint string indicating the
+--                        item shape. Plugins reimplementing `vim.ui.select` may
+--                        wish to use this to infer the structure or semantics of
+--                        `items`, or the context in which select() was called.
+--                      • match_on_detail (boolean|nil) optional flag to use the detail when filtering the items. items must have a detail field.
+--       • {on_choice}  (function) ((item|nil, idx|nil) -> ()) Called once the
+--                      user made a choice. `idx` is the 1-based index of `item`
+--                      within `items`. `nil` if the user aborted the dialog.
+
+local function vscode_ui_select(items, opts, on_choice)
+  -- fix-up defaults
+  items = items or { 'a', 'b', 'c' }
+  opts = opts or {}
+  opts.prompt = vim.F.if_nil(opts.prompt, "Select one of")
+  opts.format_item = vim.F.if_nil(opts.format_item, tostring)
+  opts.kind = vim.F.if_nil(opts.prompt, "string")
+  opts.match_on_detail = vim.F.if_nil(opts.match_on_detail, false)
+  on_choice = vim.schedule_wrap(vim.F.if_nil(on_choice, function(_, _) end))
+
+  -- fill the vscode datastructures
+  local vscode_items = {}
+  for _, item in ipairs(items) do
+    table.insert(vscode_items, {
+      label = item.label or item,
+      detail = item.detail or nil,
+      description = item.description or nil,
+    })
+  end
+  local vscode_opts = {
+    title = opts.prompt,
+    placeHolder = opts.prompt,
+    matchOnDetail = false,
+    matchOnDescription = false,
+  }
+
+  -- open the command palette
+  vscode.action("ui_select", {
+    args = {
+      items = vscode_items,
+      opts = vscode_opts,
+    },
+    callback = function(err, res)
+      if not err then
+        on_choice(res, 1)
+      else
+        on_choice(nil, -1)
```

Comment:
I believe the logic is incorrect.

https://github.com/neovim/neovim/blob/8f08b1efbd096850c04c2e8e2890d993bd4d9f95/runtime/lua/vim/ui.lua#L37-L38